### DISCUSSION TITLE
Shared coordination layer for collaborative agent runs (SETI@home-style)

### DISCUSSION THREAD

**Discussion by heidiEC** [karpathy/autoresearch#72] [Ideas]
@karpathy you mentioned wanting to make autoresearch "asynchronously massively collaborative for agents (SETI@home style)." I built the coordination layer, a shared causal graph that agents read before running and write to after.

## What it does

Every agent that runs autoresearch connects to the same shared 'pod'. Before running, it reads what the community already knows. After running, it writes back what it found. Certainty scores rise as independent agents on different hardware confirm the same results. Contradictions get flagged automatically.

**Current state of the graph** (seeded from Discussion sessions #32 and #43):

```
═══════════════════════════════════════════════════════════
  Pod Summary — 23 claims from 2 sessions
═══════════════════════════════════════════════════════════

PATTERNS
  cross-run-reproduction-fragility  value:99% Z=0.29
  fixed-budget-bitter-lesson-inversion  value:94% Z=0.29
  popular-defaults-dont-transfer  value:84% Z=0.29

TOP CLAIMS
  throughput-over-params: Z=0.46 (highest certainty — confirmed across runs)
  interaction-effects-untested: Z=0.33
  label-smoothing-catastrophic: Z=0.30

  Events: 23 │ Avg Certainty: 0.30 │ Need Validation: 23
═══════════════════════════════════════════════════════════
```

The knowledge graph with relationships between findings:

```mermaid
graph TD
    classDef pattern fill:#e8f5e9,stroke:#1b5e20
    classDef claim fill:#f3e5f5,stroke:#6a1b9a

    throughput[throughput-over-params<br/>v:100% Z:0.46]:::claim
    bitter[fixed-budget-bitter-lesson-inversion<br/>v:94% Z:0.29]:::pattern
    defaults[popular-defaults-dont-transfer<br/>v:84% Z:0.29]:::pattern
    fragility[cross-run-reproduction-fragility<br/>v:99% Z:0.29]:::pattern
    swiglu[swiglu-popularity-vs-performance<br/>v:75% Z:0.30]:::claim
    label[label-smoothing-catastrophic<br/>v:95% Z:0.30]:::claim
    interaction[interaction-effects-untested<br/>v:69% Z:0.33]:::claim
    depth[depth-width-razor-edge<br/>v:84% Z:0.29]:::claim
    sliding[sliding-window-sweet-spot<br/>v:98% Z:0.28]:::claim

    throughput ==>|ENABLES| bitter
    throughput -.->|RELATED_TO| depth
    swiglu -->|BUILDS_ON| defaults
    label -->|BUILDS_ON| defaults
    depth -->|BUILDS_ON| interaction
    fragility -->|BUILDS_ON| interaction
```

## How it works

The system uses [MCP](https://modelcontextprotocol.io/) (Model Context Protocol) — an open standard that works with Claude Code, Claude Desktop, Cursor, Cline, Gemini CLI, and any MCP-compatible client. Agents authenticate with a GitHub PAT and get scoped access to the shared pod.

Every agent follows the same loop:

```
READ  →  check what the community knows, find gaps
RUN   →  run autoresearch as normal
WRITE →  encode findings as claims with calibrated certainty
POST  →  summarize your run in a Discussion
```

The pod uses an adaptive learning rate where low-certainty claims learn fast from new evidence, high-certainty claims resist noise. When 5 independent agents on different GPUs all confirm that batch halving helps, `throughput-over-params` climbs toward consensus. When someone gets the opposite result on different hardware, it gets flagged as a potential hardware-specific divergence rather than blindly overwriting.

**Gap analysis** tells agents what's most valuable to test next:

```
🔴 High Priority
  sliding-window-sweet-spot: v=0.98 Z=0.28  — single source, needs replication
  cross-run-reproduction-fragility: v=0.99 Z=0.29  — single source, needs replication
  rope-base-frequency-200k: v=0.93 Z=0.28  — single source, needs replication
```

## Why this architecture

This isn't just a shared database. The learning system underneath is adapted from the Nervous Machine Framework, a causal inference engine originally built for physical edge devices — hardware nodes that run prediction-error loops locally without needing inference, then sync findings when connected.

The core primitive is the same:

- **Dissipative learning**: η(Z) = 1 / (1 + e^(10*(Z - 0.5))) — the sigmoid learning rate means low-certainty claims move fast on new evidence, high-certainty claims resist noise. This is how the graph self-stabilizes without a central authority.
- **Vector causal weights**: In the physical system, edges carry multi-dimensional weights (strength, confidence, context, stability). The agent-facing version simplifies this to value + certainty, but the update mechanics are identical.
- **Error signal → weight update**: ε = |predicted - actual| drives learning. In the physical system, sensors provide ground truth. Here, independent agent runs provide the validation signal. Same math, different data source.
- **Bradford Hill validation**: The full system uses 9 rigorous causal criteria (strength, consistency, specificity, temporality, dose-response, plausibility, coherence, experiment, analogy). The agent pod implements a subset through cross-run validation, contradiction detection, and gap analysis.

The goal is decentralized models that learn without centralized coordination — whether those learners are IoT sensors, edge devices, or AI agents running experiments. The autoresearch pod is the same architecture applied to a community of agents doing hyperparameter research.

What makes this work for SETI@home-style collaboration:

- **No coordinator needed** — agents read/write independently, the learning function handles convergence
- **Hardware-aware** — an H100 finding and an RTX 5070 finding can both contribute with provenance, linked as RELATED_TO rather than overwriting each other
- **GNN-ready** — the graph structure supports feature extraction for pattern refinement (finding meta-patterns across claims that individual runs can't see)
- **Scales horizontally** — each agent run is a validation event. More agents = faster convergence, not more contention

### 1. Get access

```bash
curl -X POST https://context.nervousmachine.com/auth/github/token \
  -H "Content-Type: application/json" \
  -d '{"github_token": "YOUR_GITHUB_PAT"}'
```

Returns a scoped access token. Your PAT is validated once against GitHub and never stored.

### 2. Connect your MCP client

```json
{
  "mcpServers": {
    "autoresearch-pod": {
      "command": "npx",
      "args": [
        "mcp-remote",
        "https://context.nervousmachine.com/mcp",
        "--header",
        "Authorization: Bearer YOUR_ACCESS_TOKEN"
      ]
    }
  }
}
```

### 3. Read → Run → Write

```
get_pod_summary(user_id: "autoresearch-mar5")      # what do we know?
get_gap_analysis(user_id: "autoresearch-mar5")      # what needs testing?
get_curiosity_triggers(user_id: "autoresearch-mar5") # what should I investigate?
```

After your run:

```
save_event(...)      # new finding
apply_learning(...)  # confirm or contradict existing claim
link_events(...)     # connect related findings
```

The full SKILL.md with calibration tables, value scales, and error magnitude reference is here: [SKILL.md](https://github.com/heidiEC/autoresearch-agent/blob/main/SKILL.md)

## What this enables

- **No duplicate work** — agents check what's already known before running
- **Cross-hardware validation** — findings from H100s, RTX 4090s, RTX 5070s all contribute with provenance
- **Automatic prioritization** — gap analysis surfaces the highest-value experiments to run next
- **Contradiction detection** — when results conflict, the graph flags it instead of silently overwriting
- **Convergence tracking** — watch certainty climb as independent agents replicate findings

23 claims are already in the graph from sessions #32 and #43. Every new agent run makes the shared knowledge more reliable.

---

*Built with [Event Pod](https://context.nervousmachine.com) by [Nervous Machine](https://nervousmachine.com) — dissipative learning for decentralized causal graphs.*

**Comment by aniruddhaadak80** [comment]
A SETI@home-style distributed agent swarm is such an exciting concept. If we could pool hundreds of community GPUs to have different agents exploring parallel branches of the hyperparameter space, the search tree would collapse incredibly fast.

The main challenge will definitely be syncing the `results.tsv` and making sure agents aren't duplicating the exact same idea/commits. I'd love to see a central lightweight server or Redis queue manage the global state. Such a cool long-term direction for the project.

**Reply by heidiEC** [reply]
I spun up a visualization tool for the graph. Now your agent can input your findings directly to MCP with the [SKILL.md](https://github.com/heidiEC/autoresearch-agent/blob/main/SKILL.md) and you can see the visualization at: ide.nervousmachine.com using your github credentials to login. But also, you can just configure the MCP server for any agent and just ask questions about all the findings.
<img width="1507" height="793" alt="Screenshot 2026-03-10 at 9 53 20 PM" src="https://github.com/user-attachments/assets/bdc4b044-6e01-4ae4-830f-6a46c17370b6" />

**Comment by Nyrok** [comment]
The "no duplicate work" property is the most valuable one here. The gap analysis directing agents toward high-value unexplored space is exactly the kind of coordination that makes the search collapse faster.

One thing I'd add: the input-side complement matters too. If each agent run uses inconsistently structured prompts (varying role framing, output format spec, constraint framing), the findings come back in formats that are harder to merge into the shared graph reliably. The pod solves the shared state side, but consistent structured input is what makes each agent's output parseable and comparable across runs.

I've been building [flompt](https://flompt.dev) for the input side, a visual prompt builder that makes context, objective, constraints, and output format distinct typed blocks. Open-source: github.com/Nyrok/flompt
